active/0002-centalized-mgmt-resource-types-provider-versions.md

```diff
@@ -0,0 +1,262 @@
+- With [Radius](https://github.com/radius-project/radius), it's conceivable that we may want users to **not** have `az` imported by default.
+
+### Proposed Changes
+
+- [Updating `bicepconfig.json` merge semantics](#updating-bicepconfigjson-merge-semantics)
+- [Resolution of namespace identifiers by VSCode extension when inspecting published module sources](#resolution-of-namespace-identifiers-by-vscode-extension-when-inspecting-published-module-sources)
+- [Wildcard semantics for provider versions in `bicepconfig.json`](#wildcard-semantics-for-provider-versions-in-bicepconfigjson)
+
+## Detailed design
+
+### Client side changes
+
+#### Add a new syntax for the provider import declaration statements: `import {providerName} [as {optionalAlias}]`
+
+The example below shows a usage of the proposed syntax
+
+```bicep
+provider az
+provider kubernetes as k8s // uses optional aliasing of the
+```
+
+The resolution of the symbols will be resolved from the builtin `bicepconfig.json` under a new section:
+
+```json
+{
+  // prior bicepconfig.json sections
+  // ...
+
+  "providers": {
+    "az": {
+      // A package reference string (below) gestures Bicep to source the types from the registry address.
+      "source": {
+        "registry": "mcr.microsoft.com/bicep/providers",
+        "version": "0.2.3",
+      },
+      "isDefaultImport": true // Gestures Bicep to implicitly import the namespace identifier to the global scope
+    },
+    "kubernetes": {
+      // Using the value "builtin" (below) gestures Bicep to source types from the statically compiled NuGet package reference
+      "source": {
+        "builtin": true
+      },
+      "isDefaultImport": false
+    }
+  }
+}
+```
+
+Since previous provider declaration syntax forms continue to be supported, we introduce the following constraints:
+
+- To ensure consistency with pre-existing handling of `bicepconfig.json` the configuration file closest to the Bicep file in the directory hierarchy is used.
+- The `sys` namespace is coupled to the Bicep bits version and cannot be overriden or dynamically uploaded using the mechanism described above
+
+#### Remove support for syntax `import '{providerName}@{providerVersion}' [as {optionalAlias}]`
+
+The syntax below currently in use to import built-in providers will become invalid and result in a diagnostic as a result of this change
+
+```bicep
+provider 'sys@1.0.0'
+provider 'kubernetes@1.0.0' as k8s
+```
+
+#### Change the delimiter '@' used by the supported provider declaration syntax to ':' to align with module reference syntax and increase consistency
+
+The syntax shown below uses ':' instead of '@' as a delimiter for provider declaration statements
+
+```bicep
+provider 'br:mcr.microsoft.com/bicep/providers/az:0.2.3' // notice its NOT using az@0.2.3
+provider 'br/public:az:0.2.3' as myAz
+```
```

Comment:
Is this section still relevant? Two sections up, the proposal is to move registry and versioning info from `provider` statements to bicepconfig

Reply:
It is still relevant since the syntax will continue to be accepted. The addition of centralized provider version management is not exclusive with the syntax highlighted above.

Reply:
... is the plan to keep both all the way through GA of this feature? [You make some pretty convincing arguments above as to why we shouldn't allow direct registry references inline](https://github.com/Azure/bicep-reps/pull/2/files#diff-1c3e322d927a938d42051e08eb8e935b46bd1fbfc6360cc11cb63ecab3df2914R48), and supporting both inline and configuration-driven provider sourcing seems like it would result in the maximum possible complexity (both in terms of code complexity and in terms of the complexity for users adopting this feature).

Reply:
The current plan of record is to support all three mechanisms of specifying a provider. The conversation on whether this should be the only and exclusive way to do it is not in scope for this proposal but it is a valid question and we can have a separate discussion about it. @anthony-c-martin, can you please add your thoughts on this?

Reply:
FWIW, I would be ok if we started with only the centralized method via bicepconfig and then expanding as needed/requested.

Reply:
+1. I think maintaining multiple methods of specifying provider versions and later removing certain mechanisms could result in breaking changes.

Reply:
This discussion on whether or not we want an in-line mechanism feels orthogonal to this spec. Could we agree to what @asilverman is proposing (change the delimiter) and plan to have a follow-up discussion on whether or not we should support in-line overrides generally?

Reply:
Fair... I had just assumed this section had just been accidentally left in from a previous version since my impression from the team discussion of this proposal was that the inline mechanism would be replaced. But that does seem like something we can deal with while this proposal is still active.